Support creation and deletion of groups

draffensperger opened the pull request:
This allows creation of new Google groups, and it also implements the other HTTP methods for the api.

aliang commented:
Some test broke I think because of the API change in the oauth gem. I should probably fix the version requirement

aliang commented:
I still need to review this, my fault. Supposedly there was going to be time during the holiday season, lol

draffensperger commented:
Yeah, no worries. I need to get it to pass Travis, and I think ideally we would merge the refactoring of the specs first, then I could rebase on that.

Jeiwan commented on lib/google_contacts_api/group.rb:
```diff
@@ -40,5 +40,18 @@ def edit_link
+        File.new(File.dirname(__FILE__) + '/templates/group.xml.erb').read
+
+      # Passes in group_attrs to the template via binding
+      ERB.new(@@new_group_template).result(binding)
```
I think this is a bad approach. Every time you send a POST request you have to read a file from the disk, and this is a slow operation. Actually, you can build this XML in memory using Gyoku gem (https://github.com/savonrb/gyoku).

aliang replied:
From what I can tell, that's not actually true, the result of the file read is stored in memory because of the class variable `@@new_group_template`

aliang replied:
That having been said, it might be better to use a heredoc to load the template instead of doing a file read.

Jeiwan replied:
Yep, you're right, it's stored in the class variable. But still, having a separate template is not correct. I think the whole group creation should be considered with the contact creation: these are actually two very similar features, that share the same logic. The only differences between them are the url and parameters. That's why it's not good to store a template for groups. It's just a one-time solution. Instead, we need to think about a mechanism that accepts parameters for a new contact or group and then generates an XML using these parameters. We need to have one mechanism to create entities, that will be shared by contact and group creation.
If you agree with this I'll later create an issue with general design of such mechanism.

aliang replied:
I do agree in principle and would welcome the PR, but I would like to avoid having another dependency on an XML gem. For now the template solution is ok as long as stuff is escaped properly, it's not like you can't refactor later.

draffensperger replied:
One thing I learned from doing group and contact creation in production is that there are some very small, subtle and annoying differences in what the XML should be depending on whether it's a contact create or update (and presumably that may apply to group create or update). For reference, here's the template code I'm using in my fork of the gem for the contact creation: https://github.com/CruGlobal/google_contacts_api/blob/master/lib/google_contacts_api/templates/contact.xml.erb.

The enclosing tag should be `atom:entry` for create and just `entry` for update. It's possible we could just use one or the other, but I do remember facing a confusing bug because the Google contacts API expected it in a specific format.

So while I definitely think that having a general XML generation strategy would be nice, I think that if we first approach it on a case by case basis using templates and we have good specs and a way to test things against the real Google contacts API, then if we can refactor in a clean way then we should do that eventually. But yeah, my thought is that templates are a good short/medium term approach.

Jeiwan replied:
Ok, let it be templates. Maybe we won't need to change this. But if I have time I'll anyway try to think about how to build this XML generator.

draffensperger replied:
Yeah, I'm definitely open to doing an XML generator. I think if we get the create group and create/update contact logic in with templates and good specs for them, then building an XML generator would be just a matter of refactoring the code which uses the same specs.

draffensperger commented:
I rebased this on the wonderfully modular spec refactor by @Jeiwan and also made the group creation XML based on an interpolated string rather than an erb file. I also added a group deletion method which can be helpful when testing this against the real api (create a group and delete it).

I tried both calls against the live Google api and they worked.

aliang commented on lib/google_contacts_api/group.rb:
```diff
@@ -40,5 +40,25 @@ def edit_link
+    end
+
+    def self.xml_for_create(title)
+      '<atom:entry xmlns:gd="http://schemas.google.com/g/2005" '\
```
I'd still use a heredoc here. You can still interpolate in the heredoc, too:

http://jeff.dallien.net/posts/optional-behavior-for-ruby-heredocs

draffensperger replied:
The reason I liked it as a regular string is that the resulting XML won't have any extra white space in it, but we can indent it in a reasonable way within the code block. But yeah, I'll change it to a heredoc.

draffensperger commented:
These now uses a heredoc and I also moved the create/delete methods to the `Groups` module which I think makes them easier to use since you don't need to pass in and `api` but can rather just do something like `user.create_group('title')` or `user.delete_group(group)` which parallels the existing get call which you do as `user.groups`.

For the spec, I also pulled in the `equivalent-xml` gem which makes matching XML less painful. It may seem like overkill for this particular example, but when I developed my forked version it was a huge help once you get into complex XML for creating/updating contacts and especially batched updates, so I think it makes sense to bring it in now so that we have an example of how to use it.

draffensperger commented:
@aliang does this look good, or any suggestions / requests to change?

aliang commented:
I just got back from holiday, will look tonight or tomorrow night

aliang commented on lib/google_contacts_api/groups.rb:
```diff
@@ -23,5 +23,28 @@ def get_groups(params = {})
       GoogleContactsApi::GroupSet.new(response.body, @api)
     end
+
+    def create_group(title)
```
Nothing to fix here, but this made me realize I probably shouldn't have implemented this as a module, but maybe as a "fetcher" object outside of the user/group/contact hierarchy.

aliang replied:
Also, eventually we might want to support more than just setting the title, but it's a good first step.

draffensperger replied:
Yeah, it looks like you can have arbitrary "extended properties" for Google groups (https://developers.google.com/google-apps/contacts/v3/#creating_contact_groups) but I didn't notice any other core attributes you can set besides `title` though I may have missed them and giving the ability to add extended attributes to groups could be helpful in the future for some people.

aliang commented on lib/google_contacts_api/groups.rb:
```diff
@@ -23,5 +23,28 @@ def get_groups(params = {})
+      Group.new(Hashie::Mash.new(JSON.parse(response.body)).entry)
+    end
+
+    def delete_group(group)
```
Should this be implemented on the group itself?

draffensperger replied:
Yeah, implementing it on the group does make sense and just letting it take an id would make it more generally usable. Perhaps I can move (or alias) it next time I work on the gem.

aliang commented:
Sorry it took so long to look at this, I got sick shortly after returning from holiday.

It's :+1: with comments

aliang commented on lib/google_contacts_api/result.rb:
```diff
@@ -22,6 +21,12 @@ def id
     end
 
+    # Item path relative to Google contacts URL base for puts/posts/deletes
+    def id_path
```
Does it make more sense to hard-code a URL template of some sort here?

draffensperger replied:
Good question. I think perhaps as we eventually have more Group/Contact mutation functionality (create/edit/delete) then it may become more clear what the role of this method is and where it should go. I think in my fork of the gem I use it for both contacts and groups since they follow a similar URL naming scheme in that it should have `full` in the URL for creates/updates but the "get" calls return the URL with `base`.
